### ISSUE TITLE
[Bug]: Two NC instances on one server interfere with each other - disabling apcu helps

### ISSUE THREAD

**Issue by aadlung** [nextcloud/server#37610]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

I updated from nextcloud 25 to 26, and I have 2 instances of nextcloud running on the same server with differnet domains, databases and data folders. The open_basedir and permission structure does not allow them to see/read each other.

Nevertheless, after upgrading I could see errors in nextcloud.log that one instance tries to access files of the oter, and after restarting php-fpm and accessing the second instance, it tried to access files of the other as well - only the first instance I opened kept working, the second showed a HTTP500 and the errors in log with file access.

I have `"memcache.local": "\\OC\\Memcache\\Redis",` enabled with a Redis configuration using a separate `dbindex` for each of the instances.

Only after disabling `apcu` it worked. Even if the memcache config should use Redis with its own database.

### Steps to reproduce

1. Have 2 instances on the same system
2. Upgrade from NC 25 to NC 26
3. Enable apcu

### Expected behavior

Two instances should not interfere with each other, even with apcu they should use an apcuPrefix or similar. I have no idea if this is related to the upgrade or anything else - but that's the only action I executed.

### Installation method

Community Manual installation with Archive

### Nextcloud Server version

26

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.1

### Web server

Nginx

### Database engine version

MySQL

### Is this bug present after an update or on a fresh install?

Updated to a major version (ex. 22.2.3 to 23.0.1)
--> Also tested with a FRESH install, same behavior after adding memcached+redis config with active apcu module

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "simpleSignUpLink.shown": false,
        "trusted_domains": [
            "files.example.com"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/files.example.com",
        "dbtype": "mysql",
        "version": "26.0.0.11",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "theme": "",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "dbindex": 2,
            "port": 6379
        },
        "updater.release.channel": "stable",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "loglevel": 2,
        "default_phone_region": "AT",
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - activity: 2.18.0
  - circles: 26.0.0
  - cloud_federation_api: 1.9.0
  - comments: 1.16.0
  - contactsinteraction: 1.7.0
  - dav: 1.25.0
  - federatedfilesharing: 1.16.0
  - federation: 1.16.0
  - files: 1.21.1
  - files_pdfviewer: 2.7.0
  - files_rightclick: 1.5.0
  - files_sharing: 1.18.0
  - files_trashbin: 1.16.0
  - files_versions: 1.19.1
  - firstrunwizard: 2.15.0
  - logreader: 2.11.0
  - lookup_server_connector: 1.14.0
  - nextcloud_announcements: 1.15.0
  - notifications: 2.14.0
  - oauth2: 1.14.0
  - password_policy: 1.16.0
  - photos: 2.2.0
  - privacy: 1.10.0
  - provisioning_api: 1.16.0
  - recommendations: 1.5.0
  - related_resources: 1.1.0-alpha1
  - serverinfo: 1.16.0
  - settings: 1.8.0
  - sharebymail: 1.16.0
  - support: 1.9.0
  - survey_client: 1.14.0
  - systemtags: 1.16.0
  - text: 3.7.2
  - theming: 2.1.1
  - twofactor_backupcodes: 1.15.0
  - updatenotification: 1.16.0
  - user_status: 1.6.0
  - viewer: 1.10.0
  - weather_status: 1.6.0
  - workflowengine: 2.8.0
Disabled:
  - admin_audit: 1.16.0
  - bruteforcesettings: 2.6.0
  - dashboard: 7.6.0 (installed 7.0.0)
  - encryption: 2.14.0
  - files_external: 1.18.0
  - suspicious_login: 4.4.0
  - twofactor_totp: 8.0.0-alpha.0
  - user_ldap: 1.16.0
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs
Tested with a FRESH install. It worked, until I added the memcached config with Redis

```
'memcache.local' => '\\OC\\Memcache\\Redis',
  'redis' =>
  array (
    'host' => 'localhost',
    'dbindex' => 2,
    'port' => 6379,
  ),

```

```
{"reqId":"dqf7hKO1FtScvE769qZz","level":3,"time":"2023-04-06T08:20:11+00:00","remoteAddr":"127.0.0.1","user":"webapp2","app":"PHP","method":"GET","url":"/apps/files/?dir=/&fileid=2","message":"include(): open_basedir restriction in effect. File(/var/www/webapp1/apps/firstrunwizard/lib//AppInfo/Application.php) is not within the allowed path(s): /var/www/webapp2/:/usr/lib/php/:/var/lib/php/sessions/:/usr/share/GeoIP/:/tmp/:/opt/nextcloud-data-webapp2) at /var/www/webapp2/lib/composer/composer/ClassLoader.php#578","userAgent":"Mozilla/5.0 (X11; Linux aarch64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36","version":"26.0.0.11","data":{"app":"PHP"}}
```

**Comment by toxpal** [comment]
This is unbelievable!

After reading your post, I thought "this is crazy, how is it even possible for one NC instance to access files from other instance?" and now I have the same problem. If the same server runs two or more NC26 installations, only one of them works. What's even more crazy is that each installation contains tons of errors about OTHER installation files (for example, if 1st instance is running at **/home/user1** and other instance is running in **/home/user2**, 1st instance **nextcloud.log** file contains errors with **/home/user2** path). And some of instances stop working randomly, resulting in 500 internal error.

I used NC for years and never had the same problem. WTF??

**Comment by aadlung** [comment]
I assume that just the cache is not using a corresponding prefix or similar - with Redis you can provide a database at least but for caching there should be ideally at least the instance ID or similar which is unique.

**Comment by toxpal** [comment]
I don't have Redis installed, I use apcu, and never had such problem. Only after upgrading to NC26 it started returning 500 internal server error, and it happens because each installation somehow tries to access files from other installation - https://help.nextcloud.com/t/nc-25-0-5-to-nc-26-internal-server-error/159632/4

**Comment by markushagge** [comment]
I can confirm the problem. Here also some Nextcloud installations after update to Nextcloud26 suddenly with error 500 no longer accessible, in the nextcloud log are then various error messages regarding include errors with paths to completely different installations.

Disabling APCu then brought relief.

Doesn't anyone at Nextcloud test new releases on a system running multiple Nextcloud instances?

**Comment by Bewo** [comment]
I can confirm the problem as well. I manage several servers, both with Nextcloud and with other online products. All instances always cleanly separated in their own PHP sockets, etc. I have never seen a problem of this kind. I haven't bothered to analyze the cache entries yet, but I can confirm that disabling APCu works. I currently use Redis for the local cache as well. I haven't done a benchmark yet, but noticeably there is no noticeable performance degradation.
So, that's a workaround / solution for me.

**Comment by szaimen** [comment]
Looks like https://github.com/krakjoe/apcu/issues/238

**Comment by Bewo** [comment]
> Looks like [krakjoe/apcu#238](https://github.com/krakjoe/apcu/issues/238)

I can not confirm this, but i don't think so.
It worked fine for years before NC 26. I had separate apcu-caches for each instance using apcu.

**Comment by markushagge** [comment]
Here the same. PHP-FPM is used, but it has worked for several years with various versions of Nextcloud and APCu without problems

**Comment by Haio11** [comment]
Same here after update from 25 to 26.
Two webhosted Nextclouds on the same server.
Using different PHP versions (8.1 and 8.2) solves the problem for me.

Support says: The problem is probably that APCu builds up a cache in the working memory which cannot be identified correctly. As a result, one Nextcloud tries to access data from the other.

**Comment by toxpal** [comment]
> Support says: The problem is probably that APCu builds up a cache in the working memory which cannot be identified correctly

How it comes everything works fine on NC25 and problem happens only after upgrading NC, not apcu?

**Comment by blizzz** [comment]
A prefix is used across all caches, and it's a hash of concatenated values of the instanceId, the NC version and the webroot of the instance. But it does not matter for APCu when you have not have it configured. APCu is then not used by Nextcloud itself.

While I did not observe strange behaviour (might be luck) i could find that composer was caching autoloaded classes without a path. We never configured composer to act like this however. 

If you like to doublecheck on your instance, make this a script like this available via web (ideally not public; CLI won't work);

```php
print("<pre>");
print_r(apcu_fetch('composer_autoloadOCA\Activity\AppInfo\Application'));
print("<br/>");
print_r(apcu_fetch('composer_autoloadOCA\GroupFolders\Listeners\LoadAdditionalScriptsListener'));
print("<p></p>");
print_r(apcu_cache_info());
```

Replace the keys in apcu_fetch commands with ones you see in the output below from apcu_cache_info().

**Comment by eehmke** [comment]
I installed that script in my test instance, and called it on the web. In fact, the output starts with:

```
/var/www/nextcloud-eehmke/apps/activity/lib//AppInfo/Application.php

Array
(
    [num_slots] => 4099
    [ttl] => 0
    [num_hits] => 2014292
    [num_misses] => 26156
    [num_inserts] => 26160
    [num_entries] => 1000
    [expunges] => 0
    [start_time] => 1681811184
    [mem_size] => 642000
    [memory_type] => mmap
    [cache_list] => Array
    ...

```
    where nextcloud-eehmke is my working instance. In the test instance there is no apcu activated. How to proceed?

**Comment by blizzz** [comment]
My output is:

![Screenshot_20230420_230134](https://user-images.githubusercontent.com/2184312/233486979-96f9c74c-310a-40a8-9263-08b4ae8aedc1.png)


I found two keys referencing classes for different instances. Composer will make load that class that is already cached in apcu. Meanwhile, we found the culprit.

**Comment by eehmke** [comment]
Sorry, being not an apcu expert, what would be the next step? Can I somehow clear the apcu cache? Still, why does this affect an instance that does not use apcu?

**Comment by blizzz** [comment]
Please try whether  https://github.com/nextcloud/server/pull/37853 fixes the issue for you.

**Comment by blizzz** [comment]
> Sorry, being not an apcu expert, what would be the next step? Can I somehow clear the apcu cache? Still, why does this affect an instance that does not use apcu?

It's a third party component using it, when available.

**Comment by berny156** [comment]
> Please try whether #37853 fixes the issue for you.

Works for me! Thanks a lot for the fix!

**Comment by eehmke** [comment]
Will check again after the weekend. Other priorities now.

**Comment by eehmke** [comment]
Ok I applied the patch in lib/base.php in my test instance, and the solved the problem indeed. Thank you.

**Comment by toxpal** [comment]
Can anyone confirm if this is fixed in 26.0.2? I tried upgrading to 26.0.1 earlier this week and the issue still existed in 26.0.1.

**Comment by eehmke** [comment]
@toxpal Yes I can confirm the patch is contained in 26.0.2. I just did the upgrade and I checked that.
I ran 26.0.1 for some weeks now with the patch applied, and had no issues.

**Comment by toxpal** [comment]
Thanks for the info. If I understand right, you applied patch to 26.0.1 manually (as it wasn't the part of release)?

**Comment by eehmke** [comment]
@toxpal That's right, I edited the file lib/base.php and changed that line.
When you upgrade to 26.0.2, you must not do anything.

**Comment by toxpal** [comment]
Upgraded to 26.0.2 today and can confirm the issue is fixed.

**Comment by aadlung** [comment]
Can confirm it's working - I've checked the redis cache as well, and no wrong paths are now in there.

**Comment by melroy89** [comment]
Since NC32.0 this might became an issue again.. Or is it just me? I suddenly see a web page like with an "App update required" page from another NC instance (I notice that from the footer, and in the title bar)!?

<img width="1532" height="903" alt="Image" src="https://github.com/user-attachments/assets/b31e929b-02ae-414a-afa0-7ee664182879" />

Also the App update message itself is not correct. Since all apps are updated as well.. So something there is a problem with the cache anyway.. Restarting PHP-FPM didn't resolved it (I was hoping OPCache clearing would fixed it, but it doesn't)..

**Comment by fysch** [comment]
> Since NC32.0 this might became an issue again.. Or is it just me?

Can confirm that.

I have multiple instances running on my server and whenever I try to open pages from two instances at the same time, one results in an error page showing the Nextcloud name of the other one. Disabling APCu resolves the issue.

I just discovered the issue, but have not yet looked into it in detail.

**Comment by melroy89** [comment]
> > Since NC32.0 this might became an issue again.. Or is it just me?
> 
> Can confirm that.
> 
> I have multiple instances running on my server and whenever I try to open pages from two instances at the same time, one results in an error page showing the Nextcloud name of the other one. Disabling APCu resolves the issue.
> 
> I just discovered the issue, but have not yet looked into it in detail.

Hopefully this pr will fix the regression again.. Again. https://github.com/nextcloud/server/pull/55383

**Comment by blizzz** [comment]
Could be a regression. But please open a new issue, as it has been two years since.

UPDATE: the relevant code has not changed since.

**Comment by melroy89** [comment]
I already mentioned the PR which solves the issue (see my previous comment), the regression was here: https://github.com/nextcloud/server/pull/55383/files

**Comment by blizzz** [comment]
Sorry for not reading carefully.
